src/azul/openapi.py

```diff
@@ -2,23 +2,65 @@
 from typing import List
 
 from azul.deployment import aws
-from azul.types import JSON
+from azul.types import (
+    JSON,
+    MutableJSON,
+)
 
 
-def openapi_spec(description: JSON):
+def openapi_spec(path: str, methods: List[str], path_spec: MutableJSON, method_spec: MutableJSON):
     """
-    Add description to decorated function
+    Add description to decorated function at specified route
 
-    >>> @openapi_spec({'a': 'b'})
+    >>> @openapi_spec('/foo', ['GET'], {'a': 'b'}, {'c': 'd'})
     ... def foo():
     ...     pass
+
+    >>> foo.path_spec
+    {'/foo': {'a': 'b'}}
+
+    >>> foo.method_spec
+    {('/foo', 'get'): {'c': 'd'}}
+
+    Multiple routes can be made for the same function
+
+    >>> @openapi_spec('/foo', ['GET', 'PUT'], {'a': 'b'}, {'c': 'd'})
+    ... @openapi_spec('/foo/too', ['GET'], {}, {'e': 'f'})
+    ... def foo():
+    ...     pass
+
+    >>> foo.path_spec
+    {'/foo': {'a': 'b'}}
+
+    >>> foo.method_spec
+    {('/foo/too', 'get'): {'e': 'f'}, ('/foo', 'get'): {'c': 'd'}, ('/foo', 'put'): {'c': 'd'}}
+
+    Only one route should define the top level spec
+
+    >>> @openapi_spec('/bar', ['PUT'], {'bad, duplicate': 'method spec'}, {'e': 'f'})
+    ... @openapi_spec('/bar', ['GET'], {'a': 'b'}, {'c': 'd'})
+    ... def bar():
+    ...     pass
+    Traceback (most recent call last):
```

Comment:
What about 

```
@openapi_spec('/bar', ['PUT'], {…}, {…})
def foo():
    pass
@openapi_spec('/bar', ['GET'], {…}, {…})
def bar():
    pass
```

Should that not raise an exception, too?

Reply:
This is an important error case to check for, but it can't be caught here. It's needs to be tested for in `join_specs`.